kratos/spatial_containers/bins_dynamic_objects.h

```diff
@@ -808,96 +821,49 @@ virtual void SearchObjectsInRadiusExclusive(IteratorType const& ThisObjects, Siz
 //************************************************************************
 //************************************************************************
 
-    virtual void CalculateCellSize()
+    void CalculateCellSize(std::size_t ApproximatedSize, const std::size_t GridSize[3])
     {
-
-        CoordinateType delta[Dimension];
-        CoordinateType alpha[Dimension];
-        CoordinateType mult_delta = 1.00;
-        SizeType index = 0;
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            delta[i] = mMaxPoint[i] - mMinPoint[i];
-            if ( delta[i] > delta[index] )
-                index = i;
-            delta[i] = (delta[i] == 0.00) ? 1.00 : delta[i];
-        }
-
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            alpha[i] = delta[i] / delta[index];
-            mult_delta *= alpha[i];
+        std::size_t average_number_of_cells = static_cast< std::size_t >(
+            std::pow(static_cast<double>(ApproximatedSize), 1.00 / Dimension)
+        );
+        
+        std::array<double, 3 > lengths;
+        double average_length = 0.00;
+
+        for(int i = 0; i < Dimension; i++) {
+            lengths[i] = mMaxPoint[i] - mMinPoint[i];
+            average_length += lengths[i];
         }
 
+        average_length *= 1.00 / 3.00;
 
-        mN[index] = static_cast<SizeType>( pow(static_cast<CoordinateType>(mObjectsSize/mult_delta), 1.00/Dimension) +1 );
-
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            if(i!=index)
-            {
-                mN[i] = static_cast<SizeType>(alpha[i] * mN[index]);
-                mN[i] = ( mN[i] == 0 ) ? 1 : mN[i];
+        if((GridSize[0] == 0) || (GridSize[1] == 0) || (GridSize[2] == 0)) {
```

Comment:
This still prevents fixing only 1 dimension of the bins. I think we agree with @RiccardoRossi yesterday that it would be interesting if that was an option. I can change it but I don't know if it still makes sense after using this new version of the code.